zinc/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala

```diff
@@ -54,33 +51,42 @@ final class MixedAnalyzingCompiler(
       callback: XAnalysisCallback,
       classfileManager: XClassFileManager
   ): Unit = {
+    val output = config.currentSetup.output
     val outputDirs = outputDirectories(output)
     outputDirs.foreach { d =>
-      if (!d.getPath.endsWith(".jar"))
+      if (d.getName.endsWith(".jar"))
+        IO.createDirectory(d.getParentFile)
+      else {
         IO.createDirectory(d)
+      }
     }
 
-    val incSrc = sources.filter(include)
+    val incSrc = config.sources.filter(include)
     val (javaSrcs, scalaSrcs) = incSrc.partition(javaOnly)
     logInputs(log, javaSrcs.size, scalaSrcs.size, outputDirs)
 
     /** Compile Scala sources. */
     def compileScala(): Unit =
       if (scalaSrcs.nonEmpty) {
-        val sources = if (order == Mixed) incSrc else scalaSrcs
-        val arguments = cArgs(Nil, absClasspath, None, options.scalacOptions)
-        timed("Scala compilation", log) {
-          compiler.compile(
-            sources.toArray,
-            changes,
-            arguments.toArray,
-            output,
-            callback,
-            reporter,
-            config.cache,
-            log,
-            progress.toOptional
-          )
+        JarUtils.withPreviousJar(output, callback) { extraClasspath =>
```

Comment:
I know Zinc is full of these closures that take `compile` as a function and wrap it as a function, but for the sake of maintainability and the sake of reusability I would prefer to have the logic of `withPreviousJar` be split into two methods: the first part that sets up the extra classpath and the second part that merges jars and makes sure the state is consistent. After the logic is split, we can invoke the methods directly (wrapping them up in try/catch as necessary in the use site -- as this is the only use site).

Reply:
I am not sure about this change. Maybe it can be somehow improved, but I ended up with such code:

```scala
    def compileScala(): Unit = {
      def compile(extraClasspath: Seq[File] = Nil): Unit = {
        val sources = if (config.currentSetup.order == Mixed) incSrc else scalaSrcs
        val cp = toAbsolute(extraClasspath) ++ absClasspath
        val arguments = cArgs(Nil, cp, None, config.currentSetup.options.scalacOptions)
        timed("Scala compilation", log) {
          config.compiler.compile(
            sources.toArray,
            changes,
            arguments.toArray,
            output,
            callback,
            config.reporter,
            config.cache,
            log,
            config.progress.toOptional
          )
        }
      }

      if (scalaSrcs.nonEmpty) {
        JarUtils.preparePreviousJar(output, callback) match {
          case Some(prevJar) =>
            val outputJar = JarUtils.getOutputJar(output).get
            try {
              compile(Seq(prevJar))
            } catch {
              case e: Exception =>
                IO.move(prevJar, outputJar)
                throw e
            }
            JarUtils.cleanup(outputJar, prevJar)
          case None => compile()
        }
      }
    }
```

```scala
  def preparePreviousJar(output: Output, callback: AnalysisCallback): Option[File] = {
    for {
      outputJar <- getOutputJar(output)
      prevJar <- toOption(callback.previousJar())
    } yield {
      IO.move(outputJar, prevJar)
      prevJar
    }
  }

  def cleanup(outputJar: File, prevJar: File): Unit = {
    if (outputJar.exists()) {
      JarUtils.mergeJars(into = prevJar, from = outputJar)
    }
    IO.move(prevJar, outputJar)
  }
```

It doesn't look like an improvement to me. Not only is the part of compiling with previous jar logic hidden in analysis callback `previousJar` method, but now it is also in MixedAnalyzingCompiler and 2 public methods in JarUtils. Should preparing classpath involve moving the jar? No idea, but most likely the `cleanup` method should. So it sounds like prev jar moving logic should be in jarUtils, but we still need to have part of moving logic in MixedAnalyzingCompiler as we catch the exception there. 

I don't feel like it got any better, originally, at least almost all the logic for compiling with prev jar was in one place which feels to be more maintainable.

Reply:
> Should preparing classpath involve moving the jar?
  
IMO, yes, and the more explicit it is the btter.
  
I think that this code is an improvement over the previous code because it's more explicit. I propose that you submit it. If you want to make it as similar as the previous code, you can create your `withPreviousJar` clojure that handles the setup and the cleanup with a try catch. To clarify why this code is better: it's more extensible as it allows me to call `preparePreviousJar` and `cleanup` in build tools wrapping zinc without requiring to duplicate all the logic.

Reply:
I hope I understood this correctly. I left `withPreviousJar` as it was from the outside. Inside I extracted 2 methods `preparePreviousJar` and `cleanupPreviousJar` and made them public. I guess thanks to this we have both readability, ~easy reasoning about the code and the extensibility you mention.

From minor changes to what I posted above, I return both prev jar and output jar from `prepare..` as it was quite awkward to extract it multiple times, someone might wonder if it is safe, etc.
I also removed the type param from withPreviousJar, as we don't need to catch any result, just simplified the code a bit.

On side note, I am not sure why would you need to prepare prev jar when you are wrapping zinc that already would prepare the prev jar. Also would you need to handle a case when you only prepare prev jar but never cleanup? That would not be possible with just `withPreviousJar` but doesn't make much sense to me. Other than that you can put any function that will do the compilation given extra classpath. I am just curious on specific case, but you can ignore it.